source/Halibut/Transport/PollingClient.cs

```diff
@@ -13,47 +15,66 @@ public class PollingClient : IPollingClient
+        Thread? pollingClientLoopThread;
         bool working;
-        readonly Func<RetryPolicy> createRetryPolicy;
-
-        [Obsolete("Use the overload that provides a logger. This remains for backwards compatibility.")]
-        public PollingClient(Uri subscription, ISecureClient secureClient, Func<RequestMessage, ResponseMessage> handleIncomingRequest, Func<RetryPolicy> createRetryPolicy)
-            : this(subscription, secureClient, handleIncomingRequest, null, createRetryPolicy)
-        {
-        }
 
-        public PollingClient(Uri subscription, ISecureClient secureClient, Func<RequestMessage, ResponseMessage> handleIncomingRequest, ILog log, Func<RetryPolicy> createRetryPolicy)
-            : this(subscription, secureClient, handleIncomingRequest, log, CancellationToken.None, createRetryPolicy)
-        {
-        }
+        Task? pollingClientLoopTask;
+        readonly CancellationTokenSource workingCancellationTokenSource;
+        readonly CancellationToken cancellationToken;
+        
+        readonly Func<RetryPolicy> createRetryPolicy;
+        readonly AsyncHalibutFeature asyncHalibutFeature;
+        
 
-        public PollingClient(Uri subscription, ISecureClient secureClient, Func<RequestMessage, ResponseMessage> handleIncomingRequest, ILog log, CancellationToken cancellationToken, Func<RetryPolicy> createRetryPolicy)
+        public PollingClient(Uri subscription, ISecureClient secureClient, Func<RequestMessage, ResponseMessage> handleIncomingRequest, ILog log, CancellationToken cancellationToken, Func<RetryPolicy> createRetryPolicy, AsyncHalibutFeature asyncHalibutFeature)
         {
             this.subscription = subscription;
             this.secureClient = secureClient;
             this.handleIncomingRequest = handleIncomingRequest;
             this.log = log;
             this.cancellationToken = cancellationToken;
+            workingCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
             this.createRetryPolicy = createRetryPolicy;
-            thread = new Thread(ExecutePollingLoop);
-            thread.Name = "Polling client for " + secureClient.ServiceEndpoint + " for subscription " + subscription;
-            thread.IsBackground = true;
+            this.asyncHalibutFeature = asyncHalibutFeature;
+            
         }
 
         public void Start()
         {
             working = true;
-            thread.Start();
+
+            if (asyncHalibutFeature == AsyncHalibutFeature.Disabled)
+            {
+                pollingClientLoopThread = new Thread(ExecutePollingLoop!);
+                pollingClientLoopThread.Name = "Polling client for " + secureClient.ServiceEndpoint + " for subscription " + subscription;
+                pollingClientLoopThread.IsBackground = true;
+            }
+            else
+            {
+                pollingClientLoopTask = Task.Run(async () => await ExecutePollingLoopAsyncCatchingExceptions());
```

Comment:
Do we really want to change it from a Thread to a Task?

Reply:
I am confused, is this not what making it async means?

Reply:
I think the answer is yes.

Otherwise we end up with `new Thread(() => ExecutePollingLoopAsyncCatchingExceptions().GetAwaiter().GetResult())`;

So we end up making a thread we don't need.

Doesn't matter too much in tentacle, however it will be less efficient when we have many clients e.g. in tentacle simulator or tests within Halibut.

Reply:
Just wondering, how long running is this thread?

I.e., do we want it taking a thread from the thread pool, and it'll finish with it quickly? Or should we mark it as a long running task, and let it devote a thread to it properly etc.?